text/0000-glimmer-components.md

```diff
@@ -0,0 +1,1386 @@
+side effects. For example, let's say we have a component that has a `fullName`
+computed property and expects `firstName` and `lastName` arguments:
+
+```js
+// components/person.js
+export default Component.extend({
+  firstName: null,
+  lastName: null,
+
+  fullName: computed('firstName', 'lastName', function() {
+    return `${this.firstName} ${this.lastName}`;
+  })
+});
+```
+
+The public API of this component is _supposed_ to just be those two arguements.
+However, a developer may realize that they can pass `fullName` directly to the
+component, overriding the computed property:
+
+```hbs
+<Person @fullName={{this.fullNameWithMiddle}}>
+```
+
+This is clearly a bad pattern, but it shows that in effect that most details of
+a component's implementation are not truly private, and any property or value
+can be overriden from external contexts. In most common day-to-day scenarios
+developers just have to be careful that they are following the intended public
+API of a component, but this also has the potential for misuse and enables
+antipatterns.
+
+Glimmer components assign their arguments to the `args` property on their
+instance, preventing namespace collisions from happening in the first place.
+This allows component authors to define a clear public API for a component which
+_cannot_ be circumvented.
+
+### Immutable Arguments
+
+In classic components argument values on the component are also _mutable_. This
+can lead to some confusing behavior, because argument values in the _class_ can
+change, but named arguments in _templates_ cannot. For instance, given this
+component:
+
+```js
+// components/welcome.js
+export default Component.extend({
+  firstName: 'Jen',
+  lastName: 'Weber'
+});
+```
+```hbs
+<!-- templates/components/welcome.hbs  -->
+Hey there, {{@firstName}} {{@lastName}}!
+```
+
+And this invocation:
+
+```hbs
+<Welcome />
+```
+
+You might expect that the result would be:
+
+```hbs
+Hey there, Jen Weber!
+```
+
+However, `{{@firstName}}` and `{{@lastName}}` would actually be empty values.
+They refer _directly_ to the arguments passed into the _invocation_ of the
+component, and to get the results we wanted, we would have to invoke it like so:
+
+```hbs
+<Welcome @firstName="Jen" @lastName="Weber" />
+```
+
+The advantage in this is that named arguments are fully transparent. When seen
+in a template, users can know without a doubt that the named argument was a
+value passed in from the invocation. Likewise, when they see a standard binding
+to a value like `{{this.firstName}}` or `{{firstName}}`, they know this is a
+value defined on the component - it could a computed property, it could come
+from a service, it could be a random value, but it is _not_ an argument.
+
+Glimmer components align their argument access with named args by making
+arguments available exclusively on a (shallow) frozen object, `this.args`.
+Attempting to modify `this.args` will hard error, meaning that like templates,
```

Comment:
Apologies for this question may be too small. What's 

> Attempting to modify `this.args` will hard error, ....

I guess you are trying to say something like 'will throw error' right?

Reply:
Yes, exactly!